rfc/2020/conformance-validation/mayankshah1607/rfc.md

```diff
@@ -0,0 +1,443 @@
+2. The test suite (the `_test.go` files)
+
+To run the test suite, users will be required to execute `./run.sh` with a bunch of flags (more details below). Further, the output of the tests shall be written to a filepath passed as a flag to `./run.sh`.
+
+### 2. Using Sonobuoy
+
+Sonobuoy is a great tool when it comes to conformance validation. Its [plugin model](https://sonobuoy.io/docs/master/plugins/) can allow us to easily extend functionality beyond K8s conformance validation. Essentially, Sonobuoy runs your test suite in a pod in the cluster, hence, also saves developers the overhead of configuring the desired objects such as Namespaces, ServieAccounts, ClusterRoleBindings, Jobs, RBAC, etc.
+
+Now, for the ease of management of the project, we simply extend our default approach as described above, to provide a [sonobuoy plugin](https://sonobuoy.io/docs/master/plugins/). Hence, we require 2 new items :
+
+**1. A Docker image to house our test suite**
+
+As per Sonobuoy's requirements, the test suite and its dependencies shall be packaged as a Docker container. This container shall run as a job in the cluster. In the Dockerfile, we define that the following must happen
+
+- Setup and install all the binaries/tools required by the testing environment - `linkerd`, `kubectl`, `mysql`, etc.
+- Copy _run.sh_ file that initiates the tests. (Note that we are reusing the same `run.sh` file from the default approach).
+- Copy the test suite
+- ENTRYPOINT `./run.sh`
+
+
+> Note: When generating a Sonobuoy plugin, the plugin definition file sets the start-up command as `./run.sh` by default. If we set the container’s ENTRYPOINT as `go test`, this may get overridden to `./run.sh` because of how Sonobuoy generates and runs plugins. The generated plugin definition file may however be modified to not override the startup command of the container, but this may have a negative impact on the user experience. It is important that we maintain consistency of usage in this aspect.
+
+We may also have to host this Docker image on a public repository, for E.g - _gcr.io/linkerd.io/conformance:v1.0.0_. The significance of this can be understood from the next step.
+
+**2. A sonobuoy plugin file (YAML)**
+
+The main function of Sonobuoy is running plugins; each plugin may run tests or gather data in the cluster. A plugin is defined using a YAML file that holds information such as 
+
+- Name of plugin
+- Format of result
+- Type of a plugin - Job or DaemonSet
+- Command to execute
+- args/flags for the the command
+- The Docker image to use
+- volumeMounts - for storing results
+
+We shall provide a plugin file, _l5d_conformance.yaml_ for running our test framework as a Sonobuoy plugin. It may look something like :
+```yaml
+sonobuoy-config:
+  driver: Job
+  plugin-name: l5d-conformacne
+  result-format: raw
+spec:
+  args:
+  - --enableTest="TestAutoProxyInjection/*" \
+  - --failFast \
+  - --ha \
+  - --addon-config=addon-config.yaml
```

Comment:
When doing the conformance in this way, How are users going to overwrite `addon-config`, as it would be looking inside the running job container for the file, right?

Reply:
Yes, this would actually only be an issue while running on Sonobuoy. To tackle that, I've added a new `./run.sh` flag - `--addon-config-sonobuoy`.
The RFC talks about running the tests in 2 ways:
1. Default mode (user's local environment or a container) - here, we may provide a default `addon.yaml` that the users can edit, and because its on a local environment, `linkerd install` can directly accept it. 
In case of running it in a container, the users include the `addon.yaml` file in the build process of the Dockerfile we provide.

2. Sonobuoy - We use the previously mentioned `--addon-config-sonobuoy` flag. Under the hood, it takes your local `addon.yaml` file and copies it onto the Sonobuoy pod using `kubectl cp`

Reply:
Just put the `addon.yaml` in a folder (e.g., `config`, `etc`, `data`), and update your `Dockerfile` to always `COPY` everything in that folder.